### DISCUSSION TITLE
Supporting url in ClientOptions to be a function that returns a Promise

### DISCUSSION THREAD

**Discussion by edwardfeng-db** [enisdenjo/graphql-ws#145] [General]
Hi, currently we have a use case that there is service that's checking csrf token upon the initial request, and it's expecting the csrf token to be part of the query string (url in this context). Basically we are trying to make the url follow this pattern `wss://graphql-ws?csrf=<token>`.

The csrf token could expire, so when there's an invalid token error, we would like to refetch and recreate a `graphql-ws` client with the updated token. Therefore, to make it recreate the client using the updated token, we would like to make `url` to support a function type that returns a Promise, so it can be dynamically updatable.

Specifically, we'd like our use case to be like this:
`createClient({
  url: () => new Promise((resolve) => { resolve(getUrlWithLatestCsrfToken()) }),
  lazy: true,
})`

I made a PR here such that the url now has type `string | (() => Promise<string> | string)`, which is similar to `connectionParams`. @enisdenjo Can you let me know your thought on this?

**Comment by enisdenjo** [comment]
Hi there, before we begin, this request belongs to Issues, not Discussions. Simply because it is an enhancement and you can close the Issue with a PR using the `Closes #N` syntax in the description. Discussions are merely for discussions, ideas, questions, stuff that dont require any form of modifications. Would be great if you could open an issue instead and update the PR to close it once it merges.

Nevertheless, I like the idea! Good thinking. 😄

**Reply by edwardfeng-db** [reply]
Thanks @enisdenjo! I created an issue here https://github.com/enisdenjo/graphql-ws/issues/146. Can you take another look? Thanks!

**Reply by edwardfeng-db** [reply]
Thanks so much for handling this so quickly @enisdenjo! I'm also wondering when approximately will the new change be released?

**Reply by enisdenjo** [reply]
Released in [`v4.3.0`](https://github.com/enisdenjo/graphql-ws/releases/tag/v4.3.0). 😄
